### ISSUE TITLE
Way too many types for Tokens

### ISSUE THREAD

**Issue by mystor** [dtolnay/syn#205]
Right now we have, to my current count, ~10 different types which we use to talk about a series of tokens in and around synom.

Edit: crossing off ones that have been eliminated.

* ~`syn::TokenTree`~
* ~`Vec<syn::TokenTree>`~
* `proc_macro::TokenStream`
* `proc_macro2::TokenStream`
* `proc_macro2::TokenTree`
* ~`Vec<proc_macro2::TokenTree>`~
* ~`quote::Tokens`~
* `quote::ToTokens`
* `synom::SynomBuffer`
* `synom::Cursor`

We also talk about `proc_macro2::{TokenTree, TokenStream}` as ~`synom::{TokenTree, TokenStream}`~ in some places, as they are re-exported there.

This is a complete mess, and makes it extremely hard to, for example, compose a synom parser and run it over the body of a macro which was parsed. We currently expose the following APIs for parsing:

`synom::Synom::parse()` (on types implementing `Synom` - takes a `synom::Cursor`)
`syn::parse()` (takes a `proc_macro::TokenStream`, which is useful for custom derives, but is not useful when you have any of the other types, as there is no cheap conversion from any other tokens type to TokenStream)
`syn::parse_str()` (takes a `&str`, re-parses it, and doesn't preserve span information)

Which means that if you have, say a `quote::Tokens`, you can't actually run a parser without going through some olympics to convert the types around. We really out to have a `syn::parse_tokens()` or similar which handles `quote::Tokens`.

It also would be fantastic if we could somehow get away with having a more reasonable set of token types - perhaps using `quote::Tokens` in places in the AST where we want to provide tokens, so we can have a better interface than `TokenStream` which we can control, and providing helpers for calling `Synom` parsers with `quote::Tokens`? This could at least let us get rid of `syn::TokenTree` which would be great.

@dtolnay @alexcrichton to see if they have any ideas.

**Comment by alexcrichton** [comment]
`Vec<TokenTree>` I think should always be `TokenStream`? We could also maybe delete `syn::TokenTree` in favor of `proc_macro2::TokenTree` I think as well. The `quote::Tokens` structure should probably be something like `TokenStreamBuilder` perhaps?

**Comment by mystor** [comment]
If we do decide that the type which we want people to be passing around in their programs is `proc_macro2::TokenStream`, then we should probably change the `quote!{}` macro to convert its internal `quote::Tokens` object into a `TokenStream` before producing a result. In addition I think we should probably re-export `proc_macro2::TokenStream` from `syn` somewhere, so that people can easily use it in their code.

Right now in code I write using the new syn, I find that I'm passing around `quote::Tokens` objects a lot (as they're really easy to create with `quote!{}` and work really well as a representation of a snippet of arbitrary code as they're easy to both create & use). 

The fact that we use `Vec<syn::TokenTree>` as the representation for, for example, macros: https://github.com/dtolnay/syn/blob/d5e985994ca58039ea22d6fde8acb821d9058f2f/src/mac.rs#L17, means that I have to do some gross fiddling with types whenever I see that type and and want to do things with it.

Basically I want a unified story for what type people pass around & store tokens as in their ASTs and internal macro logic, and we should provide the tools to make interacting with those types easy.

**Comment by alexcrichton** [comment]
I think we could canonicalize on `proc_macro2::TokenStream` everywhere perhaps? Including returning that from `quote!`?

**Comment by mystor** [comment]
I think we should probably have an answer for how we plan to cut down on the # of TokenStream types before we ship 0.12, so adding this to the milestone.  With the current setup this messes with consumers a bit more than I think is acceptable.

**Comment by dtolnay** [comment]
I independently came to the same conclusion as Alex -- standardize on `proc_macro2::TokenStream`, possibly with some re-exports where it makes sense.

**Comment by dtolnay** [comment]
Just mapping out how these types interact currently:

#### `proc_macro`

```rust
pub struct TokenStream { /* private */ }
pub struct TokenTree {
    pub various things
}

impl IntoIterator for proc_macro::TokenStream {
    type Item = proc_macro::TokenTree;
}

// mainly T = proc_macro::TokenTree
impl<T> FromIterator<T> for proc_macro::TokenStream
    where T: Into<proc_macro::TokenStream>;
```

#### `proc_macro2`

```rust
pub struct TokenStream(imp::TokenStream);
pub struct TokenTree {
    pub various things
}

impl IntoIterator for proc_macro2::TokenStream {
    type Item = proc_macro2::TokenTree;
}

// mainly T = proc_macro2::TokenTree
impl<T> FromIterator<T> for proc_macro2::TokenStream
    where T: Into<proc_macro2::TokenStream>;

// unstable
mod imp {
    pub struct TokenStream(proc_macro::TokenStream);
}

// stable
mod imp {
    pub struct TokenStream {
        inner: Vec<proc_macro2::TokenTree>
    }
}
```

#### `syn`

```rust
pub struct TokenTree(pub proc_macro2::TokenTree);

// also uses Vec<syn::TokenTree> inside attributes and macros
```

#### `synom`

```rust
pub struct SynomBuffer { weird private representation }
pub struct Cursor<'a> { weird private representation }

pub trait Synom {
    fn parse(synom::Cursor) -> synom::Cursor;
}

impl SynomBuffer {
    pub fn new(proc_macro2::TokenStream) -> Self;
    pub fn begin(&self) -> Cursor;
}

impl<'a> Cursor<'a> {
    pub fn token_stream(self) -> proc_macro2::TokenStream;
    pub fn token_tree(self) -> proc_macro2::TokenTree;
}
```

#### `quote`

```rust
pub struct Tokens {
    tts: Vec<proc_macro2::TokenTree>,
}

pub trait ToTokens {
    fn to_tokens(&self, &mut quote::Tokens);
}
```

**Comment by dtolnay** [comment]
One tricky difference: `syn::TokenTree` implements PartialEq and Hash while `proc_macro2::TokenTree` does not. May require handwriting some impls for AST structs.

**Comment by mystor** [comment]
We can also generate those impls with syn_codege if we don't want to handwrite them :-)

**Comment by dtolnay** [comment]
I did it manually for now in 9c76bcbe163ae7e0e1f1cccef20a9823d3d90548 but I filed #288 to follow up.

**Comment by dtolnay** [comment]
Unscheduling from 0.12 -- I made some progress and I think we are in acceptable shape for a release, but we can do more along these lines later.

**Comment by dtolnay** [comment]
With `quote!` now producing `proc_macro2::TokenStream` instead of `quote::Tokens` as of quote 0.6 I believe this is ready to close.
